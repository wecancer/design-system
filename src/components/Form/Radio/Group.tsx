import React from 'react'
import styled, {css} from 'styled-components'

import Radio from '.'

const Container = styled.div<{isRow: boolean}>`
  ${({isRow}) =>
    isRow
      ? css`
          display: block;
          & > *:not(:last-child) {
            margin-bottom: 1rem;
          }
        `
      : css`
          display: flex;
          & > *:not(:last-child) {
            margin-right: 20px;
          }
        `}
`

type Item = {
  id: string
  label: string
  isChecked: boolean
  isDisabled?: boolean
}

type ChangeParams = {
  item: Item
  index: number
  itemsUpdated: Item[]
}

export type Props = {
  onChange?(params: ChangeParams): void
  isDisabled?: boolean
  line: 'row' | 'column'
  items: Item[]
}

const RadioGroup = ({items, onChange, isDisabled, line = 'row'}: Props) => (
  <Container role="radiogroup" isRow={line === 'row'}>
    {items.map((item, index, arr) => (
      <Radio
        key={item.id}
        label={item.label}
        onChange={() =>
          typeof onChange === 'function' &&
          onChange({
            item,
            index,
            itemsUpdated: arr.map((value) =>
              value.id === item.id
                ? {...item, isChecked: !item.isChecked}
                : value,
            ),
          })
        }
        isChecked={item.isChecked}
        isDisabled={isDisabled || item.isDisabled}
      />
    ))}
  </Container>
)

export default RadioGroup
