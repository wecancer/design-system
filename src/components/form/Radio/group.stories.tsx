import React from 'react'
import {Story} from '@storybook/react'
import RadioGroup, {Props} from './group'

const Template: Story<Props> = (args) => <RadioGroup {...args} />

export const Default = Template.bind({})
Default.args = {
  items: [
    {id: '1', isChecked: false, label: 'Label 01'},
    {id: '2', isChecked: false, label: 'Label 02'},
    {id: '3', isChecked: false, label: 'Label 03'},
  ],
}

export default {
  title: 'Components/Form/Radio/Group',
  component: RadioGroup,
  argTypes: {
    line: {
      control: {
        type: 'select',
        options: ['row', 'line'],
      },
    },
  },
}
