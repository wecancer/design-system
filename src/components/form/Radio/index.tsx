import React from 'react'
import styled, { css } from 'styled-components'

const Container = styled.div`
  display: flex;
  align-items: center;
`

const Check = styled.div<{ isChecked: boolean }>`
  ${({ isChecked, theme }) => css`
    width: 32px;
    height: 32px;
    cursor: pointer;
    display: flex;
    position: relative;
    align-items: center;
    justify-content: center;
    border-radius: 50%;
    outline: none;
    transition-property: box-shadow, background-color;
    transition-duration: 250ms;
    transition-timing-function: ease;

    &[aria-disabled='true'] {
      opacity: 0.4;
      cursor: default;
    }

    ${
      isChecked
        ? css`
            color: ${theme.colors.bg};
            background-color: ${theme.colors.primary};

            &[aria-disabled='false']:hover {
              background-color: ${theme.colors.darkPrimary};
            }

            &[aria-disabled='false']:focus {
              box-shadow: 0 0 1px 6px ${theme.colors.focusPrimary};
            }

            &::before {
              content: '';
              width: 16px;
              height: 16px;
              border-radius: 50%;
              position: absolute;
              top: 50%;
              left: 50%;
              transform: translate(-50%, -50%);
              background-color: ${theme.colors.bg};
            }
          `
        : css`
            background-color: ${theme.colors.line};

            &[aria-disabled='false']:hover {
              background-color: ${theme.colors.placeholder};
            }

            &[aria-disabled='false']:focus {
              box-shadow: 0 0 1px 6px ${theme.colors.focusLine};
            }
          `
    }}
  `}
`

const Label = styled.label`
  ${({ theme }) => css`
    font-family: ${theme.font.familyRedesign};
    margin: 0 0 0 0.625rem;
    cursor: pointer;
  `};
`

export type Props = {
  id?: string
  label?: string
  onChange?(): void
  isChecked?: boolean
  isDisabled?: boolean
}

const Radio = ({
  id,
  label,
  isDisabled = false,
  isChecked = false,
  onChange,
  ...props
}: Props) => (
  <Container {...props}>
    <Check
      id={id}
      tabIndex={0}
      role="radio"
      onClick={onChange}
      onKeyDown={onChange}
      isChecked={isChecked}
      aria-checked={isChecked}
      aria-disabled={isDisabled}
    />
    {label && <Label htmlFor={id}>{label}</Label>}
  </Container>
)

export default Radio
