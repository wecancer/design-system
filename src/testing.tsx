import React from 'react'
import {render, RenderResult} from '@testing-library/react'
import WecancerProvider from './WecancerProvider'

type Props = {
  children: React.ReactNode
}

const AllTheProviders = ({children}: Props) => {
  return <WecancerProvider>{children}</WecancerProvider>
}

// eslint-disable-next-line
const appRender = (ui: React.ReactElement, options?: any): RenderResult =>
  render(ui, {wrapper: AllTheProviders, ...options})

// re-export everything
export * from '@testing-library/react'

// override render method
export {appRender as render}
