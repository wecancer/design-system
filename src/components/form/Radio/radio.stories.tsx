import React from 'react'
import { Story } from '@storybook/react'
import Radio, { Props } from '.'

const Template: Story<Props> = (args) => <Radio {...args} />

export const Default = Template.bind({})
Default.args = {
  label: 'Radio',
}

export default {
  title: 'Components/Form/Radio',
  component: Radio,
  argTypes: {
    isDisabled: {
      control: {
        type: 'boolean',
      },
    },
    isChecked: {
      control: {
        type: 'boolean',
      },
    },
    label: {
      control: {
        type: 'text',
      },
    },
  },
}
