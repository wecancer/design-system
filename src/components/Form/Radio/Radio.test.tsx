import Radio from '.'
import {render, screen, fireEvent} from '../../../testing'

describe('<Radio />', () => {
  it('should render snapshot correctly', () => {
    const {container} = render(
      <>
        <Radio />
        <Radio isChecked />
        <Radio isDisabled />
        <Radio isDisabled isChecked />
        <Radio isDisabled isChecked label="I'm checked and disabled" />
        <Radio label="I'm available" />
      </>,
    )
    expect(container).toMatchSnapshot()
  })

  it('should render radio checked', () => {
    render(<Radio isChecked />)
    const radio = screen.getByRole('radio')
    expect(radio.getAttribute('aria-checked')).toBe('true')
    expect(radio.getAttribute('aria-disabled')).toBe('false')
  })

  it('should render radio disabled', () => {
    render(<Radio isDisabled />)
    const radio = screen.getByRole('radio')
    expect(radio.getAttribute('aria-checked')).toBe('false')
    expect(radio.getAttribute('aria-disabled')).toBe('true')
  })

  it('should render radio disabled and checked', () => {
    render(<Radio isChecked isDisabled />)
    const radio = screen.getByRole('radio')
    expect(radio.getAttribute('aria-checked')).toBe('true')
    expect(radio.getAttribute('aria-disabled')).toBe('true')
  })

  it('should change checked status from click', () => {
    const handleClick = jest.fn()
    const checked = render(<Radio isChecked onChange={handleClick} />)
    const checkedComponent = checked.queryByRole('radio') as HTMLElement
    fireEvent.click(checkedComponent)
    expect(handleClick).toHaveBeenCalledTimes(1)

    const unchecked = render(<Radio isChecked={false} />)
    const uncheckedComponent = unchecked.container.querySelector('[role="radio"]') as HTMLElement
    expect(uncheckedComponent.getAttribute('aria-checked')).toBe('false')
  })
})
