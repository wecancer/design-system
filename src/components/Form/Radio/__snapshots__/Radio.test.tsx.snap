// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Radio /> should render snapshot correctly 1`] = `
<div>
  <div
    class="sc-bdvvtL gowonf"
  >
    <div
      aria-checked="false"
      aria-disabled="false"
      class="sc-gsDKAQ kAIVTa"
      role="radio"
      tabindex="0"
    />
  </div>
  <div
    class="sc-bdvvtL gowonf"
  >
    <div
      aria-checked="true"
      aria-disabled="false"
      class="sc-gsDKAQ eTNswP"
      role="radio"
      tabindex="0"
    />
  </div>
  <div
    class="sc-bdvvtL gowonf"
  >
    <div
      aria-checked="false"
      aria-disabled="true"
      class="sc-gsDKAQ kAIVTa"
      role="radio"
      tabindex="0"
    />
  </div>
  <div
    class="sc-bdvvtL gowonf"
  >
    <div
      aria-checked="true"
      aria-disabled="true"
      class="sc-gsDKAQ eTNswP"
      role="radio"
      tabindex="0"
    />
  </div>
  <div
    class="sc-bdvvtL gowonf"
  >
    <div
      aria-checked="true"
      aria-disabled="true"
      class="sc-gsDKAQ eTNswP"
      role="radio"
      tabindex="0"
    />
    <label
      class="sc-dkPtRN ebgbMx"
    >
      I'm checked and disabled
    </label>
  </div>
  <div
    class="sc-bdvvtL gowonf"
  >
    <div
      aria-checked="false"
      aria-disabled="false"
      class="sc-gsDKAQ kAIVTa"
      role="radio"
      tabindex="0"
    />
    <label
      class="sc-dkPtRN ebgbMx"
    >
      I'm available
    </label>
  </div>
</div>
`;
